### ISSUE TITLE
More dependency issues in CMakelists

### ISSUE THREAD

**Issue by marcespie** [mlpack/mlpack#1544]
found by a co-developer on OpenBSD...  custom commands should have BYPRODUCTS when required so that dependencies actually work.

(2 patches attached)
[patch2.txt](https://github.com/mlpack/mlpack/files/2501647/patch2.txt)
[patch1.txt](https://github.com/mlpack/mlpack/files/2501648/patch1.txt)

This makes ninja build break half-randomly again.

**Comment by rcurtin** [comment]
Thanks for the fixes!  The patches look good to me, but before I prepare to merge them I have to figure out what to do about CMake reverse compatibility.  Our minimum version is 2.8 but `BYPRODUCTS` appears in something like 3.3 or so.  So I just need to test that there's not a failure on 2.8---or bump the minimum version.  I'll open a PR shortly, but it may be a little bit until I am able to find the right reverse compatibility fix.

**Comment by marcespie** [comment]
On Tue, Oct 23, 2018 at 11:50:22AM -0700, Ryan Curtin wrote:
> 
>    Thanks for the fixes! The patches look good to me, but before I prepare
>    to merge them I have to figure out what to do about CMake reverse
>    compatibility. Our minimum version is 2.8 but BYPRODUCTS appears in
>    something like 3.3 or so. So I just need to test that there's not a
>    failure on 2.8---or bump the minimum version. I'll open a PR shortly,
>    but it may be a little bit until I am able to find the right reverse
>    compatibility fix.

Err, cmake 3.3 is over 3 years old by this point.

I surmise about everyone has updated already ?

**Comment by rcurtin** [comment]
It's true, but I want to make sure the software can still be built on RHEL6 and Ubuntu 14.04, since a lot of people especially at universities with slow-moving IT processes can be stuck on that.  When I checked yesterday, I saw that the cmake package in the trusty repositories was 2.8.12.2... but what I did not see until I double-checked just now is that there is a cmake3 package that is sufficiently new, and this is also the situation for EPEL6.

So, I'll just update the minimum required CMake version in the PR.  :)
